### ISSUE TITLE
Add `DCROverfittingProtection` metric

### ISSUE THREAD

**Issue by frances-h** [sdv-dev/SDMetrics#719]
### Problem Description
Currently in SDMetrics, we are offering [DisclosureProtection](https://docs.sdv.dev/sdmetrics/metrics/metrics-glossary/disclosureprotection) as the only metric for measuring data privacy. We've vetted this metric to make sure that it's understandable and insightful. However, running this metric requires the user to make some decisions about what to input as known columns and sensitive columns. This means that the privacy score is only ever based on a subset of columns.

We'd like to create a new metric that could be run on an entire set of columns.

### Expected behavior
Add the `DCROverfittingProtection` metric. This metric uses a DCR (distance to closest record) computation to measure whether the synthetic data has been overfit to the real data – as compared to a holdout set.

```python
from sdmetrics.single_table import DCROverfittingProtection

DCROverfittingProtection.compute(
  real_training_data=real_df,
  synthetic_data=synthetic_df,
  real_validation_data=holdout_df,
  metadata=my_metadata)

DCROverfittingProtection.compute_breakdown(
  real_training_data=real_df,
  synthetic_data=synthetic_df,
  real_validation_data=holdout_df,
  metadata=my_metadata)
```

#### `compute_breakdown`
__Args__
- `real_training_data (pd.DataFrame)`: A pandas.DataFrame object containing the real data used for training the synthesizer.
- `synthetic_data (pd.DataFrame)`: A pandas.DataFrame object containing the synthetic data sampled from the synthesizer.
- `real_validation_data (pd.DataFrame)`: A pandas.DataFrame object containing a holdout set of real data. This data should not have been used to train the synthesizer. For this metric to work effectively, this metric should be roughly the same size as the real data.
- `metadata (dict)`: A metadata dictionary that describes the table of data.
- `num_rows_subsample (int, optional)`: The number of synthetic data rows to subsample from the synthetic data. Use this to increase the speed of the computation, if your dataset is large. Default is `None` (do not subsample the data).
- `num_iterations (int, optional)`: The number of iterations to perform when subsampling. The final score will be the average of all iterations. Default is 1.

__Returns__
A dictionary breakdown of the score, with the following information:
- The overall score of the metric
- The percent of synthetic rows that were closer to the training set and the percentage of synthetic rows that were closer to the holdout set.
    - If multiple iterations were used, the percentages should be the average across all iterations.

Example
```python
{
  'score': 0.5,
  'synthetic_data_percentages': {
    'closer_to_training': 0.75,
    'closer_to_holdout': 0.25
  }
}
```

__Computation Algorithm__
1. Preprocessing:
  -> Only consider columns that are statistical (sdtype is numerical, datetime, categorical, or boolean).
  -> Convert all datetimes into UnixTimestamps so they can be treated as numerical values for the rest of the metric
  -> If requested by the user, subsample the synthetic data
2. Go through every row, `s` in the synthetic data and measure the following:
  a. Measure the DCR between `s` and the real data using the algorithm below. Call this distance `d(s, real)`
  b. Measure the DCR of `s`  in the holdout set using the same algorithm. Call this distance `d(s, holdout)`
  -> After Step 1 there will be a `d(s, real)` and `d(s, holdout)` for every row in the synthetic data. 
3. Calculate the percentage of rows where `d(s, real) < d(s, holdout)` (i.e the number of rows that are closer to the real dataset than the holdout dataset). Call this percentage `p`.
4. Return  `min((1-p)*2, 1)`
5. Repeat the above for `num_iterations` requested by the user

_DCR between a row and a dataset_
- Given a row `s` and a dataset `D`, this algorithm computes the distance to the closest record of `D`.
- Compute the record distance between `s` and every single row in `D` using the DCR between two rows algorithm below.
- The DCR is the minimum distance value that we compute between `s` and all rows of `D`.

_DCR between 2 rows (including null values)_
Given a row `s` and a row `d` compute distance between these rows.
- Loop through every column's value (j) in these rows and compute the distance. If these values are called `s_j` and `d_j`, then distance between them is defined based on the type of data:
  - For numerical data, `distance(s_j, d_j) = abs(s_j - d_j) / (range of column)` if both `s_j` and `d_j` are both non-null.
    - The `distance(s_j, d_j) = 0` if both `s_j` and `d_j` are null.
    - The `distance(s_j, d_j) = 1` if only one of `s_j` and `d_j` are null.
  - For categorical data, nulls are considered another category and treated like any other category value.
    - `distance(s_j, d_j) = 0 if s_j == d_j else 1`

The overall distance between rows s and d is the average of the column distances we compute

#### `compute`
The `compute` method should take the same arguments as the `compute_breakdown` method.

The `compute` method should just return the overall `score` parameter calculated by `compute_breakdown`.

### Additional context
See [doc](https://docs.google.com/document/d/1APIEvLfCCptC1bbFgF8bBRLnyAC_3y5pICb_9yCUxuA/edit?tab=t.jkxbunij8p7p#heading=h.va5ntny100zj)
